core/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java

```diff
@@ -551,4 +551,25 @@ public void testNestedMapInList() throws IOException {
                     parser.list());
         }
     }
+
+    public void testDotsInFieldNames() {
+        Map<String, Object> map = new HashMap<>();
+        map.put("a.b", 2);
+        Map<String, Object> sub = new HashMap<>();
+        sub.put("c", 3);
+        map.put("a", sub);
+        map.put("d", 5);
+
+        // dots in field names in includes
+        Map<String, Object> filtered = XContentMapValues.filter(map, new String[] {"a"}, new String[0]);
+        Map<String, Object> expected = new HashMap<>(map);
+        expected.remove("d");
+        assertEquals(expected, filtered);
+
+        // dots in field names in excludes
+        filtered = XContentMapValues.filter(map, new String[0], new String[] {"a"});
+        expected = new HashMap<>(map);
+        expected.keySet().retainAll(Collections.singleton("d"));
+        assertEquals(expected, filtered);
```

Comment:
Could you add a test with strings longer than one char? Sadly, you probably also want on that has a character that is two java `char` wide as well. I'd be good to assert everything works right when you fail half way through the word - say you try and match `cat` against the pattern `cbs`. The `c` works, the `a` would put you in the `-1` state, etc.